source/secure_tunneling.c

```diff
@@ -423,7 +438,23 @@ static int s_secure_tunneling_send_data(struct aws_secure_tunnel *secure_tunnel,
         AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Invalid Stream Id");
         return AWS_ERROR_IOTDEVICE_SECUTRE_TUNNELING_INVALID_STREAM;
     }
-    return s_secure_tunneling_send(secure_tunnel, data, DATA);
+    struct aws_byte_cursor new_data = {.ptr = data->ptr, .len = data->len};
+    while (new_data.len) {
+        size_t bytes_max = new_data.len;
+        size_t amount_to_send = bytes_max < AWS_IOT_ST_SPLIT_MESSAGE_SIZE ? bytes_max : AWS_IOT_ST_SPLIT_MESSAGE_SIZE;
+
+        struct aws_byte_cursor send_cursor = aws_byte_cursor_advance(&new_data, amount_to_send);
+        if (send_cursor.len) {
+            if (s_secure_tunneling_send(secure_tunnel, &send_cursor, DATA) != AWS_OP_SUCCESS) {
+                AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Failure writing data to out_buf");
+                return AWS_OP_ERR;
+            }
+        }
+        aws_mutex_lock(&mutex);
+        aws_condition_variable_wait(&condition_variable, &mutex);
```

Comment:
It's not safe to wait on a condition variable without a condition.  Theoretically the send could have completed (and the signal sent) before reaching here at which point this will deadlock.